### PR TITLE
Recycling of owned GraphicBuffers is now done in callback.

### PR DESCRIPTION
**Pull request by emilwestergren**
GraphicBuffers owned by Vidhance are now recycled back to the wrapper when freed inside Vidhance.

@sebastianbaginski peer review.

### CODE REVIEW

File: nexus6/VidhanceProcessor.cpp [hunk 1]
```diff
@@ -26,9 +26,14 @@ struct create_method_t CameraHALWrapper::mCreateMethod = { VidhanceProcessor::cr
...
 
 vidhance_context context = NULL;
+VideoProcessor* _this;
+void _recycleGraphicBuffer(void* buffer) {
+	_this->recycleCameraBuffer((GraphicBuffer*) buffer);
+}
 VidhanceProcessor::VidhanceProcessor(const char* cameraId) :
 		VideoProcessor(cameraId) {
 	vidhance_load();
+	_this = this;
```
Comment (sebastianbaginski) [inline nexus6/VidhanceProcessor.cpp]: Yeah I know. It wouldn't be needed if callbacks registered with `kean_draw_graphicBuffer_registerCallbacks` could take pointer to user data. This is how it is typically solved, so you could have
```C++
void _recycleGraphicBuffer(void* buffer, void* userData) {
      VideoProcessor* processor = reinterpret_cast<VideoProcessor*>(userData);
      assert(processor);
      processor->recycleCameraBuffer((GraphicBuffer*) buffer);
}

// then you pass `this` as user data
kean_draw_graphicBuffer_registerCallbacks(allocateGraphicBuffer, createGraphicBuffer, _recycleGraphicBuffer, (void*)this, lockGraphicBuffer, unlockGraphicBuffer);
```

**Comment by sebastianbaginski** [comment]
As you know now the global `_this` pointer is overwritten on every constructor call. Do you have only one `VideoProcessor``?

**Comment by emilwestergren** [comment]
Yes I was aware of that, it was a quick fix. I'll take care of it now.

File: nexus6/VidhanceProcessor.cpp (hunk 1 above)
Comment (emilwestergren) [inline nexus6/VidhanceProcessor.cpp]: I think `userData` should be a parameter in the vidhance graphicBuffer constructor though.
```c++
kean_draw_graphicBuffer_new(backend, userData, ...);
```
from ooc:
```ooc
This _free(this _backend, this _userData)
```

Comment (sebastianbaginski) [inline nexus6/VidhanceProcessor.cpp]: We just need a way to pass it to the framework. Pick the one that you think is best.

**Comment by emilwestergren** [comment]
Do you think we can merge this and fix the hack later? The fault doesn't even lie here, it's a flaw in the Vidhance API. I think this works good enough for the moment (we only have a single active VidhanceProcessor instance so it shouldn't cause any problems). @sebastianbaginski

**Comment by sebastianbaginski** [comment]
OK. Can you create an issue about the user data parameter ?

**Comment by emilwestergren** [comment]
This issue will address this: https://github.com/vidhance/android-camera-wrapper/issues/31

**Pull request merged**